File: PS Multi Tools/Classes/XOREngine.vb
Public Class XOREngine
    Public Shared Function GetXOR(inByteArray As Byte(), offsetPos As Integer, length As Integer, XORKey As Byte()) As Byte()
        If inByteArray.Length < offsetPos + length Then
            Throw New Exception("Combination of chosen offset pos. & Length goes outside of the array to be xored.")
        End If

        If (length Mod XORKey.Length) <> 0 Then
            Throw New Exception("Nr bytes to be xored isn't a mutiple of xor key length.")
        End If

        Dim pieces As Integer = length \ XORKey.Length

        Dim outByteArray As Byte() = New Byte(length - 1) {}

        For i As Integer = 0 To pieces - 1
            For pos As Integer = 0 To XORKey.Length - 1
                outByteArray((i * XORKey.Length) + pos) += inByteArray(offsetPos + (i * XORKey.Length) + pos) Xor XORKey(pos)
            Next
        Next

        Return outByteArray
    End Function
End Class

File: PSMultiTools/Classes/XOREngine.cs
using System;

namespace PSMultiTools.Classes
{
    public class XOREngine
    {

        public static byte[] GetXOR(byte[] inByteArray, int offsetPos, int length, byte[] XORKey)
        {
            if (inByteArray.Length < offsetPos + length)
            {
                throw new Exception("Combination of chosen offset pos. & Length goes outside of the array to be xored.");
            }

            if (length % XORKey.Length != 0)
            {
                throw new Exception("Nr bytes to be xored isn't a mutiple of xor key length.");
            }

            int pieces = length / XORKey.Length;

            byte[] outByteArray = new byte[length];

            for (int i = 0, loopTo = pieces - 1; i <= loopTo; i++)
            {
                for (int pos = 0, loopTo1 = XORKey.Length - 1; pos <= loopTo1; pos++)
                    outByteArray[i * XORKey.Length + pos] += (byte)(inByteArray[offsetPos + i * XORKey.Length + pos] ^ XORKey[pos]);
            }

            return outByteArray;
        }

    }
}